### ISSUE TITLE
Change number of bands from 4 to 3

### ISSUE THREAD

**Issue by linkyndy** [libvips/libvips#1330]
We are currently reading images and changing their colourspace with:

```
image = Vips::Image.new_from_file(filename)
image = image.colourspace(:srgb)
```

We do this because our third-party library needs to operate on an image that is written in memory as an "RGB byte array in raster read order". We write the image to memory with:

```
image.write_to_memory
```

This all works fine. However, there are PNGs that have 4 bands and we also have to write these to memory. How can we transform/convert such images so that they have 3 bands and are written to memory in the same way as the 3-band images (and as our library expects)?

**Comment by jcupitt** [comment]
Hello @linkyndy,

You can use `flatten` to squash the alpha out of an image. It defaults to a white background, but you can set any colour.

Something like:

```ruby
image = Vips::Image.new_from_file filename, access: :sequential
image = image.colourspace :srgb
if image.has_alpha?
    image = image.flatten background: [128, 255, 128]
end
memory = image.write_to_memory
```

To use a nasty greenish background colour.

The `access:` hint makes libvips stream the image from the file into your memory buffer. You should see a nice speedup and a useful drop in memory use.

**Comment by linkyndy** [comment]
Nice! I just saw the [daltonize example](https://github.com/libvips/ruby-vips/blob/master/example/daltonize8.rb), and I figured I can simply remove the 4th band. Would that produce similar results to your suggestion? In the end, if only RGB is needed by our library, this should theoretically work 🤔 

Or, are there other cases where the 4th band is **not** alpha, but something else, after converting to the sRGB colourspace?

About `access: :sequential`, isn't this the default, if `:access` is not provided?

**Comment by jcupitt** [comment]
Hi again,

Many PNGs put garbage in the transparent parts of the image to help compression. You need to mask this stuff out with `flatten`. Have a look at the output of `pngcrush`, for example.

`new_from_file` defaults to `:random`, since that works with everything. `:sequential` is a special case that only works if everything downstream is a simple top-to-bottom operation.

**Comment by linkyndy** [comment]
I understand. So you're saying dropping the 4th band is _not_ something to be done? 🤔 

Not sure I completely understand your statement about `:sequential`; we're dealing with a lot of formats (pretty much everything that can be categorised as images from a user perspective), and if this argument does not work with everything, I wouldn't use it. Or, are there specific cases where it should be used and where not?

**Comment by jcupitt** [comment]
You can go down to three bands, it's just that you should use `flatten` rather than simply dropping the alpha channel.

Re. `sequential`, there's a section in the docs on this:

http://libvips.github.io/libvips/API/current/How-it-opens-files.md.html

tldr: as long as you're not doing things like 90 degree rotate, `sequential` should work.

**Comment by linkyndy** [comment]
Works like a charm. Thank you!
